Closes #35: Modify/Fix Users Routes And Update User Collection Accordingly

Dhooly opened the pull request:
This is related to issue #35 

## Description / Changes Made

- Modified the user's routes to match the format and structure of the current 'users' collection in the database.

Routes fixed:
1. get_user
2. create_user
3. update_user
4. delete_user

- Delete all posts that are tied to a user once they delete their account

- Add proper HTTP response for get_user_posts where none was returned

- Remove unnecessary / old field in schema that will not be used (userID)

## Screenshots
![image](https://github.com/UNLV-CS472-672/2024-S-GROUP5-Munch/assets/85095643/a743e593-33d4-449f-8dae-236e68daa662)

![image](https://github.com/UNLV-CS472-672/2024-S-GROUP5-Munch/assets/85095643/4d6da465-cc7d-48cd-b3ac-abe6a1456ac2)

## How to Test

1. Install Postman
2. Create a collection called 'Munch'
3. Create a new request
6. Select the HTTP method for the route you want to test (GET, POST, DELETE, etc)
7. Add the proper route
8. Add required data if necessary (JSON data, Content-Type, etc)
9. Send the request
10. Check if expected response is received
11. Check if changes are reflected in the database
12. Repeat steps 3-11 for any other requests

## Checklist

- [x] I have added/updated relevant documentation, and I have followed the coding style guidelines.
- [x] I have added how to test these additions.
- [x] I have checked for any potential conflicts with other branches and fixed any merge conflicts.

Dhooly committed:
Closes #34: Modify/Fix Users Routes And Update User Collection Accordingly

- Add proper HTTP response for get_user_posts
- Modify Cloud Firestore user routes to follow new schema
- Remove old unecessary field in schema (userID)
- Use function to connect to DB for cleaner code

Dhooly committed:
Remove and create proper comment

Dhooly committed:
Apply formatting changes

Dhooly committed:
Merge remote-tracking branch 'upstream/main' into testing-v1

Dhooly committed:
Merge branch 'UNLV-CS472-672:main' into testing-v1

rparker2003 reviewed (changes requested):
In create_user(), update_user(), and delete_user(), these functions all try to reference the collection when they are being given new data which there is no reason to reference the collection. Because of this, I am getting errors in all 3 above functions. If you remove the reference to the collection, i.e. change 
`new_user_data["following"] = [
            db.collection("users").document(following)
            for following in data["following"]
]`
 to 
`new_user_data["following"] = [
            db.document(following)
            for following in data["following"]
]`

it will pass the errors and allow the routes to run

Dhooly committed:
Merge remote-tracking branch 'upstream/main' into testing-v1

Dhooly committed:
Fix calls to database in route that were giving errors

- Tweak create_user route to take in a collection id and not generate a
  random id.
- Adjust routes that were giving errors

Dhooly committed:
Merge branch 'UNLV-CS472-672:main' into testing-v1

Dhooly committed:
Merge branch 'testing-v1' of github.com:Dhooly/2024-S-GROUP5-Munch into testing-v1

Dhooly committed:
Apply formatting changes

Dhooly committed:
Merge remote-tracking branch 'upstream/main' into testing-v1

Dhooly committed:
Merge branch 'UNLV-CS472-672:main' into testing-v1

Dhooly committed:
Merge branch 'testing-v1' of github.com:Dhooly/2024-S-GROUP5-Munch into testing-v1

noodaj commented on backend/server.py:
```diff
@@ -590,15 +587,21 @@ def get_user(user_id):
 
-    # Convert the date to string
+    # Convert the bio and username to strings
     user_data["bio"] = str(user_data["bio"])
```
I don't think that these need to be strings. They should already be the right types when we pass them in.

Dhooly replied:
I will refactor and resolve this redundant conversion

noodaj commented on backend/server.py:
```diff
@@ -787,21 +801,25 @@ def delete_user(user_id):
+        # Get the user data
+        user_data = user_ref.get().to_dict()
+
+        # Delete all posts from the user
```
This does delete the user but I'm not sure if the references are still there for the other user's post that have references connected to this user.

noodaj reviewed (approved):
I think that this PR solves some of the issues that we have. I'm also happy that the route for the `/api/users` include the user_id. This is great since we can get the data for our users now instead of having just a plain route.

noodaj commented:
@rparker2003 I'm not sure if you guys want to refactor this now, but I see that the try_connect_to_db is called for every function. Can we not have this outside of the at the very top of the file and just return an error if we are not able to connect to the database.

rparker2003 commented on backend/server.py:
```diff
@@ -787,21 +801,25 @@ def delete_user(user_id):
+
+        # Delete all posts from the user
+        for post in user_data["posts"]:
+            post_ref = db.document(post.path)
```
post.path here needs to be changed to just post as this is currently giving me an error when trying to delete a user from the database.

rparker2003 replied:
I looked into this further and it seems the issue does not rely here. The issue relies with the create_user function as it is saving the data as strings instead of saving the data as references. Please modify create_user function and verify that the create_user() and update_user() functions save the data as references and not as strings.

Dhooly replied:
I've hardcoded the desired values for create_user because when a new user is created all schema should be initialized to its null equivalent

rparker2003 reviewed (changes requested):
Look at my above comment in the conversation history

Dhooly committed:
Merge branch 'main' of https://github.com/UNLV-CS472-672/2024-S-GROUP5-Munch into testing-v1

Dhooly committed:
Fix create_user route to store schema as references instead of strings

Dhooly committed:
Properly update post list in update_user route and store as reference in
database

rparker2003 commented:
It seems another issue that on a put request the data isn't actually even when the data is verified as a reference. Possible relates to the issue listed above.

Dhooly committed:
Fix create_user route again by forcing empty schema

Dhooly committed:
Remove validation check in create_user because we dont take in user JSON
anymore, we hardcode the schema

rparker2003 commented:
I have looked through your revisions and I like the changes. Everything functions as required!
